resources/docs/TUTORIAL_8_MODEL_OPTIMIZATION.md

```diff
@@ -0,0 +1,156 @@
+The parameter settings and the evaluation scores will be written to `param_selection.txt` in the result directory.
+While selecting the best parameter combination we do not store any model to disk. We also do not perform a test run
+during training, we just evaluate the model once after training on the test set for logging purpose.
+
+## Finding the best Learning Rate
+
+The learning rate is one of the most important hyper parameter and it fundamentally depends on the topology of the loss
+landscape via the architecture of your model and the training data it consumes. An optimal learning will improve your
+training speed and hopefully give more performant models. A simple technique described by Leslie Smith's
+[Cyclical Learning Rates for Training](https://arxiv.org/abs/1506.01186) paper is to train your model starting with a
+very low learning rate and increases the learning rate exponentially at every batch update of SGD. By plotting the loss
+with respect to the learning rate we will typically observe three distinct phases: for low learning rates the loss does
+not improve, an optimal learning rate range where the loss drops the steepest and the final phase where the loss
+explodes as the learning rate becomes too big. With such a plot, the optimal learning rate selection is as easy as
+picking the highest one from the optimal phase.
+
+In order to run such an experiment start with your initialized `ModelTrainer` and call `find_learning_rate()` with the
+`base_path` and the optimizer (in our case `torch.optim.adam.Adam`). Then plot the generated results via the
+`Plotter`'s `plot_learning_rate()` function and have a look at the `learning_rate.png` image to select the optimal
+learning rate:
+
+```python
+from flair.datasets import WNUT_17
+from flair.embeddings import TokenEmbeddings, WordEmbeddings, StackedEmbeddings
+from flair.trainers import ModelTrainer
+from typing import List
+from torch.optim.adam import Adam
+
+# 1. get the corpus
+corpus = WNUT_17().downsample(0.1)
+print(corpus)
+
+# 2. what tag do we want to predict?
+tag_type = 'ner'
+
+# 3. make the tag dictionary from the corpus
+tag_dictionary = corpus.make_tag_dictionary(tag_type=tag_type)
+print(tag_dictionary.idx2item)
+
+# 4. initialize embeddings
+embedding_types: List[TokenEmbeddings] = [
+    WordEmbeddings('glove'),
+]
+
+embeddings: StackedEmbeddings = StackedEmbeddings(embeddings=embedding_types)
+
+# 5. initialize sequence tagger
+from flair.models import SequenceTagger
+
+tagger: SequenceTagger = SequenceTagger(hidden_size=256,
+                                        embeddings=embeddings,
+                                        tag_dictionary=tag_dictionary,
+                                        tag_type=tag_type,
+                                        use_crf=True)
+
+# 6. initialize trainer
+trainer: ModelTrainer = ModelTrainer(tagger, corpus)
+
+# 7. find learning rate
+learning_rate_tsv = trainer.find_learning_rate('resources/taggers/example-ner', Adam)
```

Comment:
this function is broken, it should be changed to the code below, ideally as part of this PR.

```python
    def find_learning_rate(
        self,
        base_path: Union[Path, str],
        optimizer,
        file_name: str = "learning_rate.tsv",
        start_learning_rate: float = 1e-7,
        end_learning_rate: float = 10,
        iterations: int = 100,
        mini_batch_size: int = 32,
        stop_early: bool = True,
        smoothing_factor: float = 0.98,
        **kwargs,
    ) -> Path:
        best_loss = None
        moving_avg_loss = 0

        # cast string to Path
        if type(base_path) is str:
            base_path = Path(base_path)
        learning_rate_tsv = init_output_file(base_path, file_name)

        with open(learning_rate_tsv, "a") as f:
            f.write("ITERATION\tTIMESTAMP\tLEARNING_RATE\tTRAIN_LOSS\n")

        optimizer = optimizer(self.model.parameters(), lr=start_learning_rate, **kwargs)

        train_data = self.corpus.train

        scheduler = ExpAnnealLR(optimizer, end_learning_rate, iterations)

        model_state = self.model.state_dict()
        self.model.train()

        step = 0
        while step < iterations:
            batch_loader = DataLoader(train_data, batch_size=mini_batch_size, shuffle=True)
            for batch in batch_loader:
                step += 1

                # forward pass
                loss = self.model.forward_loss(batch)
                if isinstance(loss, tuple):
                    loss = loss[0]

                # update optimizer and scheduler
                optimizer.zero_grad()
                loss.backward()
                torch.nn.utils.clip_grad_norm_(self.model.parameters(), 5.0)
                optimizer.step()
                scheduler.step(step)

                learning_rate = scheduler.get_lr()[0]

                loss_item = loss.item()
                if step == 1:
                    best_loss = loss_item
                else:
                    if smoothing_factor > 0:
                        moving_avg_loss = smoothing_factor * moving_avg_loss + (1 - smoothing_factor) * loss_item
                        loss_item = moving_avg_loss / (1 - smoothing_factor ** (step + 1))
                    if loss_item < best_loss:
                        best_loss = loss

                if step > iterations:
                    break

                if stop_early and (loss_item > 4 * best_loss or torch.isnan(loss)):
                    log_line(log)
                    log.info("loss diverged - stopping early!")
                    step = iterations
                    break

                with open(str(learning_rate_tsv), "a") as f:
                    f.write(f"{step}\t{datetime.datetime.now():%H:%M:%S}\t{learning_rate}\t{loss_item}\n")

            self.model.load_state_dict(model_state)
            self.model.to(flair.device)

        log_line(log)
        log.info(f"learning rate finder finished - plot {learning_rate_tsv}")
        log_line(log)

        return Path(learning_rate_tsv)
```

Reply:
I addressed this as part of this PR, but note that I made tiny adjustment to the source code you pasted above. With the above code, we were getting the following warining from Torch:
> UserWarning: The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use `scheduler.step()` to step the scheduler.

this warning broke the CI build.

So I replaced the line `scheduler.step(step)` with  `scheduler.step()`. If my understanding is correct, this shouldn't badly affect the execution of the code. But if it does, please do let me know.